include/ginkgo/core/base/executor.hpp

```diff
@@ -386,45 +431,59 @@ private:                                                                     \
-            static auto name = [this] {                                       \
-                std::ostringstream oss;                                       \
-                oss << #_kernel << '#' << sizeof...(Args);                    \
-                return oss.str();                                             \
-            }();                                                              \
-            return name.c_str();                                              \
-        }                                                                     \
-                                                                              \
-        GKO_KERNEL_DETAIL_DEFINE_RUN_OVERLOAD(OmpExecutor, omp, _kernel);     \
-        GKO_KERNEL_DETAIL_DEFINE_RUN_OVERLOAD(CudaExecutor, cuda, _kernel);   \
-        GKO_KERNEL_DETAIL_DEFINE_RUN_OVERLOAD(HipExecutor, hip, _kernel);     \
-        GKO_KERNEL_DETAIL_DEFINE_RUN_OVERLOAD(DpcppExecutor, dpcpp, _kernel); \
-        GKO_KERNEL_DETAIL_DEFINE_RUN_OVERLOAD(ReferenceExecutor, reference,   \
-                                              _kernel);                       \
-                                                                              \
-    private:                                                                  \
-        mutable std::tuple<Args &&...> data;                                  \
-    };                                                                        \
-                                                                              \
-    template <typename... Args>                                               \
-    static _name##_operation<Args...> make_##_name(Args &&... args)           \
-    {                                                                         \
-        return _name##_operation<Args...>(std::forward<Args>(args)...);       \
-    }                                                                         \
-    static_assert(true,                                                       \
-                  "This assert is used to counter the false positive extra "  \
+#define GKO_REGISTER_OPERATION(_name, _kernel)                                 \
+    template <typename... Args>                                                \
+    auto make_##_name(Args &&...args)                                          \
+    {                                                                          \
+        return ::gko::detail::make_register_operation(                         \
+            #_name, sizeof...(Args), [&args...](auto exec) {                   \
+                if (std::is_same<                                              \
+                        decltype(exec),                                        \
+                        std::shared_ptr<const ::gko::ReferenceExecutor>>::     \
+                        value) {                                               \
```

Comment:
I see, that works because we use exec->run(kernel) which already evaluates to the correct executor type, that's a good approach.

For simplicity, how about doing something like this instead?
```suggestion
                using exec_type = std::decay_t<decltype(exec)::element_type>;
                if (std::is_same<                                              \
                        exec_type,                                        \
                        ::gko::ReferenceExecutor>::value) { \
```